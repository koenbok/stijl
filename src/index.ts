import * as React from "react";
import { css as glamor } from "glamor";

import { renderToString } from "react-dom/server";
import { renderStatic } from "glamor/server";

// TODO: Warn if there is a parent styling component through contexts
// TODO: Add cool types for media queries
// TODO: Add more psuedo css helpers

declare module "react" {
  interface HTMLAttributes<T> {
    css?: StylingRule;
  }

  interface SVGAttributes<T> {
    css?: StylingRule;
  }
}

const patchCreateElement = () => {
  const _createElement = React.createElement;

  const createElement = (...args) => {
    const props = Array.isArray(args) ? args[1] : args["props"];

    if (props && props.css && props.css instanceof StylingRule) {
      Object.assign(props, props.css.css());
      delete props["css"];
    }

    return _createElement.apply(this, args);
  };

  (React as any).createElement = createElement;
};

patchCreateElement();

export class StylingRule {
  private _css = {};

  constructor(css: React.CSSProperties) {
    this._css = css;
  }

  css() {
    return glamor(this._css) as any;
  }

  mediaQuery(query: string, css: React.CSSProperties) {
    this._css[`@media only screen and (${query})`] = css;
    return this;
  }

  hover(css: React.CSSProperties) {
    this._css[":hover"] = css;
    return this;
  }

  active(css: React.CSSProperties) {
    this._css[":active"] = css;
    return this;
  }

  placeholder(css: React.CSSProperties) {
    this._css[":placeholder"] = css;
    return this;
  }

  before(css: React.CSSProperties) {
    this._css[":before"] = css;
    return this;
  }
}

export const Style = (css: React.CSSProperties) => new StylingRule(css);

export const Styling = (props: { body: any }) => {
  let { html, css, ids } = renderStatic(() => renderToString(props.body));
  return React.createElement("style", {
    dangerouslySetInnerHTML: { __html: css }
  });
};
