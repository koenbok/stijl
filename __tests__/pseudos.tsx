import "jest";
import * as React from "react";
import * as ReactDOMServer from "react-dom/server";
import * as renderer from "react-test-renderer";
import { Style, Styling } from "../src/index";

const Template = props => {
  return (
    <html>
      <head>
        <Styling body={props.children} />
      </head>
      <body>{props.children}</body>
    </html>
  );
};

let style;
beforeEach(() => {
  style = Style({ border: "1px solid red" });
});

test(":active", () => {
  style.active({ border: "1px solid green" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test(":after", () => {
  style.after({ border: "1px solid green" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test(":before", () => {
  style.before({ border: "1px solid green" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test(":checked", () => {
  style.checked({ border: "1px solid green" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test(":disabled", () => {
  style.disabled({ border: "1px solid green" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test(":first-child", () => {
  style.firstChild({ border: "1px solid green" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test(":focus", () => {
  style.focus({ border: "1px solid green" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test(":hover", () => {
  style.hover({ border: "1px solid green" });
  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test(":last-child", () => {
  style.lastChild({ border: "1px solid green" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test(":nth-child", () => {
  style.nthChild("odd", { border: "1px solid green" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("::placeholder", () => {
  style.placeholder({ background: "green" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test(":visited", () => {
  style.visited({ background: "green" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
