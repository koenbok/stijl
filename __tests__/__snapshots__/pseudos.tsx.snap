// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`::placeholder 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-tp17oe,[data-css-tp17oe] {border: 1px solid red;}.css-tp17oe::-webkit-input-placeholder,[data-css-tp17oe]::-webkit-input-placeholder {background: green;}.css-tp17oe::-moz-placeholder,[data-css-tp17oe]::-moz-placeholder {background: green;}.css-tp17oe::-ms-input-placeholder,[data-css-tp17oe]::-ms-input-placeholder {background: green;}.css-tp17oe::placeholder,[data-css-tp17oe]::placeholder {background: green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-tp17oe=""
    />
  </body>
</html>
`;

exports[`:active 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1e0ptde,[data-css-1e0ptde] {border: 1px solid red;}.css-1e0ptde:active,[data-css-1e0ptde]:active {border: 1px solid green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-1e0ptde=""
    />
  </body>
</html>
`;

exports[`:after 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-kqzsho,[data-css-kqzsho] {border: 1px solid red;}.css-kqzsho:after,[data-css-kqzsho]:after {border: 1px solid green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-kqzsho=""
    />
  </body>
</html>
`;

exports[`:before 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1h2j4ma,[data-css-1h2j4ma] {border: 1px solid red;}.css-1h2j4ma:before,[data-css-1h2j4ma]:before {border: 1px solid green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-1h2j4ma=""
    />
  </body>
</html>
`;

exports[`:checked 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-hnm3i6,[data-css-hnm3i6] {border: 1px solid red;}.css-hnm3i6:checked,[data-css-hnm3i6]:checked {border: 1px solid green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-hnm3i6=""
    />
  </body>
</html>
`;

exports[`:disabled 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1n2pwdt,[data-css-1n2pwdt] {border: 1px solid red;}.css-1n2pwdt:disabled,[data-css-1n2pwdt]:disabled {border: 1px solid green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-1n2pwdt=""
    />
  </body>
</html>
`;

exports[`:first-child 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-3l785a,[data-css-3l785a] {border: 1px solid red;}.css-3l785a:first-child,[data-css-3l785a]:first-child {border: 1px solid green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-3l785a=""
    />
  </body>
</html>
`;

exports[`:focus 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-ravbph,[data-css-ravbph] {border: 1px solid red;}.css-ravbph:focus,[data-css-ravbph]:focus {border: 1px solid green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-ravbph=""
    />
  </body>
</html>
`;

exports[`:hover 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1iz9137,[data-css-1iz9137] {border: 1px solid red;}.css-1iz9137:hover,[data-css-1iz9137]:hover {border: 1px solid green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-1iz9137=""
    />
  </body>
</html>
`;

exports[`:last-child 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-tvf5a9,[data-css-tvf5a9] {border: 1px solid red;}.css-tvf5a9:last-child,[data-css-tvf5a9]:last-child {border: 1px solid green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-tvf5a9=""
    />
  </body>
</html>
`;

exports[`:nth-child 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-glqlp1,[data-css-glqlp1] {border: 1px solid red;}.css-glqlp1:nth-child(odd),[data-css-glqlp1]:nth-child(odd) {border: 1px solid green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-glqlp1=""
    />
  </body>
</html>
`;

exports[`:visited 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-z6sgur,[data-css-z6sgur] {border: 1px solid red;}.css-z6sgur:visited,[data-css-z6sgur]:visited {background: green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-z6sgur=""
    />
  </body>
</html>
`;
